storage/integration_test.go

```diff
@@ -3233,7 +3215,100 @@ func TestIntegration_WriterAppend(t *testing.T) {
 				h.mustWrite(w, tc.content)
 
 				// Download content again and validate.
-				// Disabled due to b/408373388; unskip after this is resolved.
+				// Disabled due to b/395944605; unskip after this is resolved.
+				// gotBytes := h.mustRead(obj)
+				// if !bytes.Equal(gotBytes, tc.content) {
+				// 	t.Errorf("content mismatch: got %v bytes, want %v bytes", len(gotBytes), len(tc.content))
+				// }
+
+				// Check object exists and Finalized attribute set as expected.
+				attrs := h.mustObjectAttrs(obj)
+				if tc.finalize && attrs.Finalized.IsZero() {
+					t.Errorf("got unfinalized object, want finalized")
+				}
+				if !tc.finalize && !attrs.Finalized.IsZero() {
+					t.Errorf("got object finalized at %v, want unfinalized", attrs.Finalized)
+				}
+
+			})
+		}
+	})
+}
+
+// Writer test for append takeover of unfinalized object.
+func TestIntegration_WriterAppendTakeover(t *testing.T) {
+	t.Skip("b/402283880")
+	ctx := skipAllButBidi(context.Background(), "ZB test")
+	multiTransportTest(ctx, t, func(t *testing.T, ctx context.Context, _, prefix string, client *Client) {
+		h := testHelper{t}
+		bucketName := prefix + uidSpace.New()
+		bkt := client.Bucket(bucketName)
+
+		h.mustCreateZonalBucket(bkt, testutil.ProjID())
+		defer h.mustDeleteBucket(bkt)
+
+		testCases := []struct {
+			name           string
+			finalize       bool
+			content        []byte
+			chunkSize      int
+			takeoverOffset int64
+		}{
+			{
+				name:           "first message takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: MiB,
+			},
+			{
+				name:           "first chunk takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: 3 * MiB,
+			},
+			{
+				name:           "middle chunk takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: 6 * MiB,
+			},
+			{
+				name:           "final chunk takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: 8*MiB + 100,
+			},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				// Create non-finalized appendable writer and upload first part of content.
+				obj := bkt.Object(tc.name + uidSpace.New()).Retryer(WithPolicy(RetryAlways))
+				defer h.mustDeleteObject(obj)
+				w := obj.If(Conditions{DoesNotExist: true}).NewWriter(ctx)
+				w.Append = true
+				w.FinalizeOnClose = false
+				w.ChunkSize = tc.chunkSize
+
+				h.mustWrite(w, tc.content[:tc.takeoverOffset])
```

Comment:
What happens if the first writer isn't closed (but is appropriately flushed to GCS through Flush or regular buffer flow), and we try to take over? I assume it will work? But if we try to use this first writer again after taking over, what error do we get?

Reply:
I think this should cause an error in the first writer on Close; I can test it out.

Reply:
Did you test this?

Reply:
Yes, we get a precondition failed error: 

```close write object "first message takeover-20250415-2531366975931-0004" to bucket "golang-grpc-test-20250415-2531366975931-0003": rpc error: code = FailedPrecondition desc = A different writer has become the exclusive writer of this object.```